### ISSUE TITLE
SDL OpenGL3 issue with multi windows

### ISSUE THREAD

**Issue by Vin789** [ocornut/imgui#5874]
**Version/Branch of Dear ImGui:**

Version: Dear ImGui 1.89 WIP (18828)
Branch: (https://github.com/ocornut/imgui/tree/docking)

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_sdl + imgui_impl_opengl3
Compiler: MSVC (IDE: Visual Studio 2017 (v141))
Operating System: Window 10

**My Issue/Question:**

Hi / Salut !

I was testing the code under https://github.com/ocornut/imgui/blob/docking/examples/example_sdl_opengl3/main.cpp

Following https://github.com/ocornut/imgui/issues/4264#issuecomment-868472207 I was able to draw a SDL image under ImGui (screen 1).

Adding  SDL_SetHint(SDL_HINT_RENDER_DRIVER, "opengl"); at the top of the main was necessary to draw the image. Unfortunately this call seems to have an issue with multi window. Nothing is rendered (input works) if the window is independent (screen 2).  

If I comment SDL_SetHint(SDL_HINT_RENDER_DRIVER, "opengl"); I loose the SDL image but independent window works fine (screen 3).

The full code (altered main example) is available at the end of this post. I also try to change some stuff about the SDL_Renderer creation without success. My OpenGL is really rusty (> 10 years) and I'm a SDL noob so I'm sorry in advance, I may have miss something obvious or maybe multi windows is not supported or required more code to works with SDL standard draw.

Any help is appreciated. Thanks you.

**Screenshots/Video**

![screen1](https://user-images.githubusercontent.com/13332569/200607453-01f54cba-b475-4f94-8d62-8a501b46d889.png)
![screen2](https://user-images.githubusercontent.com/13332569/200607457-25319084-9e24-4a43-a3e3-3d9bf5ec2ca8.png)
![screen3](https://user-images.githubusercontent.com/13332569/200607458-e09acdee-593a-453b-a847-3ffb48d99381.png)

**Standalone, minimal, complete and verifiable example:** _(see https://github.com/ocornut/imgui/issues/2261)_

// Dear ImGui: standalone example application for SDL2 + OpenGL
// (SDL is a cross-platform general purpose library for handling windows, inputs, OpenGL/Vulkan/Metal graphics context creation, etc.)
// If you are new to Dear ImGui, read documentation from the docs/ folder + read the top of imgui.cpp.
// Read online: https://github.com/ocornut/imgui/tree/master/docs

```
#include "imgui.h"
#include "imgui_impl_sdl.h"
#include "imgui_impl_opengl3.h"
#include <stdio.h>
#include <SDL.h>
#if defined(IMGUI_IMPL_OPENGL_ES2)
#include <SDL_opengles2.h>
#else
#include <SDL_opengl.h>
#endif

// Main code
int main(int, char**)
{
    // Seems necessary to draw SDL images but broke rendering of external window
    SDL_SetHint(SDL_HINT_RENDER_DRIVER, "opengl");

    // Setup SDL
    // (Some versions of SDL before <2.0.10 appears to have performance/stalling issues on a minority of Windows systems,
    // depending on whether SDL_INIT_GAMECONTROLLER is enabled or disabled.. updating to the latest version of SDL is recommended!)
    if (SDL_Init(SDL_INIT_VIDEO | SDL_INIT_TIMER | SDL_INIT_GAMECONTROLLER) != 0)
    {
        printf("Error: %s\n", SDL_GetError());
        return -1;
    }

    // Decide GL+GLSL versions
#if defined(IMGUI_IMPL_OPENGL_ES2)
    // GL ES 2.0 + GLSL 100
    const char* glsl_version = "#version 100";
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, 0);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_ES);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 2);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
#elif defined(__APPLE__)
    // GL 3.2 Core + GLSL 150
    const char* glsl_version = "#version 150";
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, SDL_GL_CONTEXT_FORWARD_COMPATIBLE_FLAG); // Always required on Mac
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 2);
#else
    // GL 3.0 + GLSL 130
    const char* glsl_version = "#version 130";
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, 0);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 0);
#endif

    // Create window with graphics context
    SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);
    SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 24);
    SDL_GL_SetAttribute(SDL_GL_STENCIL_SIZE, 8);
    SDL_WindowFlags window_flags = (SDL_WindowFlags)(SDL_WINDOW_OPENGL | SDL_WINDOW_RESIZABLE | SDL_WINDOW_ALLOW_HIGHDPI);
    SDL_Window* window = SDL_CreateWindow("Dear ImGui SDL2+OpenGL3 example", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, 1280, 720, window_flags);
    SDL_Renderer* renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED);
    SDL_GLContext gl_context = SDL_GL_CreateContext(window);
    SDL_GL_MakeCurrent(window, gl_context);
    SDL_GL_SetSwapInterval(1); // Enable vsync

    // Setup Dear ImGui context
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;
    io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;       // Enable Keyboard Controls
    //io.ConfigFlags |= ImGuiConfigFlags_NavEnableGamepad;      // Enable Gamepad Controls
    io.ConfigFlags |= ImGuiConfigFlags_DockingEnable;           // Enable Docking
    io.ConfigFlags |= ImGuiConfigFlags_ViewportsEnable;         // Enable Multi-Viewport / Platform Windows
    //io.ConfigViewportsNoAutoMerge = true;
    //io.ConfigViewportsNoTaskBarIcon = true;

    // Setup Dear ImGui style
    ImGui::StyleColorsDark();
    //ImGui::StyleColorsLight();

    // When viewports are enabled we tweak WindowRounding/WindowBg so platform windows can look identical to regular ones.
    ImGuiStyle& style = ImGui::GetStyle();
    if (io.ConfigFlags & ImGuiConfigFlags_ViewportsEnable)
    {
        style.WindowRounding = 0.0f;
        style.Colors[ImGuiCol_WindowBg].w = 1.0f;
    }

    // Setup Platform/Renderer backends
    ImGui_ImplSDL2_InitForOpenGL(window, gl_context);
    ImGui_ImplOpenGL3_Init(glsl_version);

    // Load Fonts
    // - If no fonts are loaded, dear imgui will use the default font. You can also load multiple fonts and use ImGui::PushFont()/PopFont() to select them.
    // - AddFontFromFileTTF() will return the ImFont* so you can store it if you need to select the font among multiple.
    // - If the file cannot be loaded, the function will return NULL. Please handle those errors in your application (e.g. use an assertion, or display an error and quit).
    // - The fonts will be rasterized at a given size (w/ oversampling) and stored into a texture when calling ImFontAtlas::Build()/GetTexDataAsXXXX(), which ImGui_ImplXXXX_NewFrame below will call.
    // - Use '#define IMGUI_ENABLE_FREETYPE' in your imconfig file to use Freetype for higher quality font rendering.
    // - Read 'docs/FONTS.md' for more instructions and details.
    // - Remember that in C/C++ if you want to include a backslash \ in a string literal you need to write a double backslash \\ !
    //io.Fonts->AddFontDefault();
    //io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\segoeui.ttf", 18.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/DroidSans.ttf", 16.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/Roboto-Medium.ttf", 16.0f);
    //io.Fonts->AddFontFromFileTTF("../../misc/fonts/Cousine-Regular.ttf", 15.0f);
    //ImFont* font = io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesJapanese());
    //IM_ASSERT(font != NULL);

    // Our state
    bool show_demo_window = true;
    bool show_another_window = false;
    ImVec4 clear_color = ImVec4(0.45f, 0.55f, 0.60f, 1.00f);

    SDL_Surface* surface = SDL_LoadBMP("D:/test.bmp");
    SDL_Texture* texture = SDL_CreateTextureFromSurface(renderer, surface);

    // Main loop
    bool done = false;
    while (!done)
    {
        // Poll and handle events (inputs, window resize, etc.)
        // You can read the io.WantCaptureMouse, io.WantCaptureKeyboard flags to tell if dear imgui wants to use your inputs.
        // - When io.WantCaptureMouse is true, do not dispatch mouse input data to your main application, or clear/overwrite your copy of the mouse data.
        // - When io.WantCaptureKeyboard is true, do not dispatch keyboard input data to your main application, or clear/overwrite your copy of the keyboard data.
        // Generally you may always pass all inputs to dear imgui, and hide them from your application based on those two flags.
        SDL_Event event;
        while (SDL_PollEvent(&event))
        {
            ImGui_ImplSDL2_ProcessEvent(&event);
            if (event.type == SDL_QUIT)
                done = true;
            if (event.type == SDL_WINDOWEVENT && event.window.event == SDL_WINDOWEVENT_CLOSE && event.window.windowID == SDL_GetWindowID(window))
                done = true;
        }

        // Start the Dear ImGui frame
        ImGui_ImplOpenGL3_NewFrame();
        ImGui_ImplSDL2_NewFrame();
        ImGui::NewFrame();

        // 1. Show the big demo window (Most of the sample code is in ImGui::ShowDemoWindow()! You can browse its code to learn more about Dear ImGui!).
        if (show_demo_window)
            ImGui::ShowDemoWindow(&show_demo_window);

        // 2. Show a simple window that we create ourselves. We use a Begin/End pair to create a named window.
        {
            static float f = 0.0f;
            static int counter = 0;

            ImGui::Begin("Hello, world!");                          // Create a window called "Hello, world!" and append into it.

            ImGui::Text("This is some useful text.");               // Display some text (you can use a format strings too)
            ImGui::Checkbox("Demo Window", &show_demo_window);      // Edit bools storing our window open/close state
            ImGui::Checkbox("Another Window", &show_another_window);

            ImGui::SliderFloat("float", &f, 0.0f, 1.0f);            // Edit 1 float using a slider from 0.0f to 1.0f
            ImGui::ColorEdit3("clear color", (float*)&clear_color); // Edit 3 floats representing a color

            if (ImGui::Button("Button"))                            // Buttons return true when clicked (most widgets return true when edited/activated)
                counter++;
            ImGui::SameLine();
            ImGui::Text("counter = %d", counter);

            ImGui::Text("Application average %.3f ms/frame (%.1f FPS)", 1000.0f / ImGui::GetIO().Framerate, ImGui::GetIO().Framerate);
            ImGui::End();
        }

        // 3. Show another simple window.
        if (show_another_window)
        {
            ImGui::Begin("Another Window", &show_another_window);   // Pass a pointer to our bool variable (the window will have a closing button that will clear the bool when clicked)
            ImGui::Text("Hello from another window!");
            if (ImGui::Button("Close Me"))
                show_another_window = false;
            ImGui::End();
        }

        // Rendering
        ImGui::Render();
        glViewport(0, 0, (int)io.DisplaySize.x, (int)io.DisplaySize.y);
        glClearColor(clear_color.x * clear_color.w, clear_color.y * clear_color.w, clear_color.z * clear_color.w, clear_color.w);
        glClear(GL_COLOR_BUFFER_BIT);

        // SDL_renderer
        SDL_Rect rct2;
        rct2.x = rct2.y = 200;
        rct2.w = rct2.h = 400;
        SDL_RenderCopy(renderer, texture, 0, &rct2);

        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());

        // Update and Render additional Platform Windows
        // (Platform functions may change the current OpenGL context, so we save/restore it to make it easier to paste this code elsewhere.
        //  For this specific demo app we could also call SDL_GL_MakeCurrent(window, gl_context) directly)
        if (io.ConfigFlags & ImGuiConfigFlags_ViewportsEnable)
        {
            SDL_Window* backup_current_window = SDL_GL_GetCurrentWindow();
            SDL_GLContext backup_current_context = SDL_GL_GetCurrentContext();
            ImGui::UpdatePlatformWindows();
            ImGui::RenderPlatformWindowsDefault();
            SDL_GL_MakeCurrent(backup_current_window, backup_current_context);
        }

        SDL_GL_SwapWindow(window);
    }

    // Cleanup
    SDL_DestroyTexture(texture);
    SDL_FreeSurface(surface);

    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplSDL2_Shutdown();
    ImGui::DestroyContext();

    SDL_GL_DeleteContext(gl_context);
    SDL_DestroyRenderer(renderer);
    SDL_DestroyWindow(window);
    SDL_Quit();

    return 0;
}
```

**Comment by PathogenDavid** [comment]
Mixing `SDL_Renderer` and OpenGL has always been problematic. (See https://github.com/ocornut/imgui/issues/1116 https://github.com/ocornut/imgui/issues/4264 https://github.com/ocornut/imgui/issues/935 https://github.com/ocornut/imgui/issues/796)

Choose one of the other, not both.

**Comment by Vin789** [comment]
@PathogenDavid Thx for your anser. When you say "Choose one of the other, not both.". Are you only talking about the SDL_renderer object or any SDL call associated to rendering in general (and so are you suggesting to move fully OpenGL for rendering and not using the SDL at all ?)

**Comment by PathogenDavid** [comment]
No problem! To me those mean the same thing.

I'm specifically saying not use [SDL's 2D accelerated rendering](http://wiki.libsdl.org/CategoryRender) at all if you want to directly use OpenGL. This basically means don't use the `SDL_Renderer` object. Using SDL's other modules (like windowing, keybaord input, and audio) is completely fine.

If you want to use `SDL_Renderer` (and never make any `gl*` calls) you can use `imgui_impl_sdlrenderer` instead of `imgui_impl_opengl3`, but do note that it doesn't support secondary viewports (which means you won't be able to pull Dear ImGui windows out of the main window.) -- I'd strongly encourage you to just use OpenGL directly though.

**Comment by Vin789** [comment]
Thx for the informations. Now I'm still thinking about all this. I want to create a 2d video game editor on top of my engine. I could go without multi-window in the first place and wait for the implementation of multi-window inside imgui_impl_sdlrenderer one day. But I'm pretty sure this is definitely not a priority inside Omar roadmap for Dear ImGui. 

So I'm evaluating the remaining options:
- Coding the support for multi-window inside imgui_impl_sdlrenderer myself (and by poking my 3D progs friends for help, going to cost me a lot of drinks and time).
- Going to DX but DX is not portable and I want to stay portable if possible. Also not sure if it mix well with SDL too.
- Going Vulkan. Could be a choice but is it possible to use SDL renderer with Vulkan ? I tried inside the sdl_vulkan example but the call to SDL_RenderCopy is not enough. And adding SDL_RenderPresent will ofc having the same flickering issue described inside the other threads with OpenGL (double render).
- Changing my whole engine to stop using SDL for drawing and go full OpenGL. I want to avoid this and keep it as it is if possible for the 2D accelerated option.

**Comment by ocornut** [comment]
Note that multi-viewport for SDL_Renderer backend is likely to be a relatively easy feature to add.

**Comment by PathogenDavid** [comment]
> * Coding the support for multi-window inside imgui_impl_sdlrenderer myself (and by poking my 3D progs friends for help, going to cost me a lot of drinks and time).

Turns out this is harder than expected, I'm going to split off my findings into a second comment.

> * Going to DX but DX is not portable and I want to stay portable if possible. Also not sure if it mix well with SDL too.

Yeah, Direct3D would likely have the same problem down the line. From what I've heard it's less problematic than OpenGL, but my understanding is `SDL_Renderer` expects to be in control of everything related to rendering.

> * Going Vulkan. Could be a choice but is it possible to use SDL renderer with Vulkan ? I tried inside the sdl_vulkan example but the call to SDL_RenderCopy is not enough. And adding SDL_RenderPresent will ofc having the same flickering issue described inside the other threads with OpenGL (double render).

I don't believe `SDL_Renderer` has a Vulkan backend at all, so this probably isn't viable either.

(Also if you're hesitant to touch OpenGL you probably shouldn't even get within spitting distance of Vulkan, IMO.)

> * Changing my whole engine to stop using SDL for drawing and go full OpenGL. I want to avoid this and keep it as it is if possible for the 2D accelerated option.

This would be my recommendation personally, but if you already have a sizable amount of `SDL_Renderer` code I can see why you wouldn't want to 😅

**Comment by PathogenDavid** [comment]
@ocornut

TL;DR: Adding multi-viewport support to `imgui_impl_sdlrenderer` requires disabling user texture support, dirty hacks, or changes to SDL.

> Note that multi-viewport for SDL_Renderer backend is likely to be a relatively easy feature to add.

So I started to look into adding this, and it turns out it actually might not be trivial. I started writing some advice to @Vin789 on how they might do it, then I decided to just do it myself, and then I found what might be a showstopper.

Each `SDL_Window` must have its own `SDL_Renderer` (it's a strict 1:1 relationship internally.) It turns out that you cannot share textures between different `SDL_Renderer` instances, which means you can't share textures between windows. [This is enforced by the API](https://github.com/libsdl-org/SDL/blob/dd72f3f03d207b23d9a7a2845950f49ded5a825f/src/render/SDL_render.c#L4172), it doesn't matter whether the render drivers match.

If you want a quick copy+pastable smoke test:

```cpp
// Add this to the initialization of example_sdl_sdlrenderer's main
SDL_Window* window2;
SDL_Renderer* renderer2;
SDL_CreateWindowAndRenderer(300, 300, window_flags, &window2, &renderer2);

// ...

// And this to the end of the main loop
SDL_SetRenderDrawColor(renderer2, 255, 0, 0, 255);
SDL_RenderClear(renderer2);
if (SDL_RenderCopy(renderer2, (SDL_Texture*)io.Fonts->TexID, nullptr, nullptr) != 0)
    SDL_Log("Error rendering texture: %s", SDL_GetError());
SDL_RenderPresent(renderer2);
```

The second window will be blank red and the log will be spammed with errors about the texture and renderer not matching.

---------------------

I see the following solutions to this problem:

* Don't allow user textures when multi-viewports is enabled and clone the font texture for each window.
  * This would be a frustrating limitation
  * Cloning the font texture not ideal for performance. Might cause noticable issues with really large CJK textures.
  * Cloned font textures has the potential to become quite annoying once the dynamic font atlas is introduced.
    * It's already pretty annoying with the public create/destroy functions exposed by the backend because it creates implementations that differ between `master` and `docking`.
* Don't allow user textures on secondary viewports and same font texture cloning.
  * This would be a confusing limitation, especially considering the audience for SDL_Renderer.
* Transparently download and reupload textures to allow sharing them between windows
  * A dirty hack
  * Would require complex caching and possibly manual invalidation for dynamic textures.
  * `SDL_Renderer` does not appear to provide the APIs necessary to do this for all types of textures. (As far as I can see there's no way to copy a GPU-only texture to a readback texture, which would be required for `SDL_TEXTUREACCESS_STATIC` and possibly `SDL_TEXTUREACCESS_TARGET` textures.)
* Enforce that users must use `SDL_SetHint(SDL_HINT_RENDER_DRIVER, "opengl");` to enable secondary viewports and render all secondary viewports using OpenGL directly.
  * Dirty hack, feels like it's asking for trouble.
  * Adds a lot of complexity to the backend.
  * Would require poking at SDL internals to get the texture's ID. (I originally thought we could use `SDL_GL_BindTexture` for this, but looking at the implementation I don't think it'd actually work because it changes the GL context to the renderer's.)
* Petition the SDL developers to add (or accept a PR to add) a new `SDL_RENDERER_TEXTURESHARING` capability flag and allow sharing textures between `SDL_Renderer` instances of the same driver when supported.
  * This is probably the correct solution but is definitely more effort than I have time for right now. (I only looked into this because I also thought it'd be relatively quick and easy.)
* Put in the effort to figure out how to properly use `imgui_impl_opengl3` with `SDL_Renderer`
  * Turns out this might be slightly more supported than previously thought (based on the presence of `SDL_GL_BindTexture`)
  * Would still have issues with user textures because `SDL_Texture` would not be usable as `ImTextureID`, user textures would need to be OpenGL textures or we'd have to poke at SDL internals as mentioned above.
* Accept the limitation and tell people not to use `SDL_Renderer` if they want multi-viewports.
